### PR TITLE
[Fixes #567] Adding port and domain parsing in HostString

### LINKED ISSUE #567: Expose port and domain in HostString
Talked with @Tratcher and he suggested this.
It's missing for some code that needs to manipulate them without implementing our own logic. c.f. https://github.com/aspnet/Mvc/pull/4113

### PR DESCRIPTION
**Pull request by sebastienros**

### CODE REVIEW

File: src/Microsoft.AspNetCore.Http.Abstractions/HostString.cs [hunk 1]
```diff
@@ -25,6 +26,31 @@ public HostString(string value)
         }
 
         /// <summary>
+        /// Creates a new HostString without modification.
+        /// </summary>
+        /// <param name="value">The value should be Unicode rather than punycode. IPv6 addresses must use square braces.</param>
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Http.Abstractions/HostString.cs]: > `value`

Really? Looks more like the `domain` from how it's handled.

> must use square braces

No "must"; the constructor adds missing square braces.

File: src/Microsoft.AspNetCore.Http.Abstractions/HostString.cs [hunk 2]
```diff
@@ -25,6 +26,31 @@ public HostString(string value)
...
+            if (value.IndexOf('[') == -1
+                && (index = value.IndexOf(':')) >= 0
+                && index < value.Length - 1
+                && value.IndexOf(':', index + 1) >= 0)
+            {
+                // IPv6 without brackets ::1 is the only type of host with 2 or more colons
+                value =  $"[{value}]";
+            }
+
+            _value = value + ":" + port;
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Http.Abstractions/HostString.cs]: Why is this done if `!port.HasValue`?

File: src/Microsoft.AspNetCore.Http.Abstractions/HostString.cs [hunk 3]
```diff
@@ -25,6 +26,31 @@ public HostString(string value)
...
         /// <summary>
+        /// Creates a new HostString without modification.
+        /// </summary>
+        /// <param name="value">The value should be Unicode rather than punycode. IPv6 addresses must use square braces.</param>
+        /// <param name="port">A positive, greater than 0 value representing the port in the host string.</param>
+        public HostString(string value, int? port)
+        {
+            if(port.HasValue && port <= 0)
+            {
+                throw new ArgumentException(Resources.FormatException_PortMustBeGreaterThanZero(nameof(port)), nameof(port));
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Http.Abstractions/HostString.cs]: Why does the resource include the parameter name? That'll be in the final `Message` regardless.

Comment (dougbu) [inline src/Microsoft.AspNetCore.Http.Abstractions/HostString.cs]: Suggest `ArgumentOutOfRangeException`.

File: src/Microsoft.AspNetCore.Http.Abstractions/HostString.cs [hunk 4]
```diff
@@ -38,6 +64,92 @@ public bool HasValue
...
+
+                // normalize the value
+                var uriComponent = ToUriComponent();
+
+                int index;
+                int port = 0;
+                if ((index = uriComponent.IndexOf(']')) >= 0)
+                {
+                    // IPv6 in brackets [::1], maybe with port
+                    if ((index = uriComponent.IndexOf(':', index)) >= 0)
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Http.Abstractions/HostString.cs]: Is it legal for the colon before the port to appear anywhere other than immediately after the closing square bracket? If not, the `uriComponent.Length == index + 1` check in `Domain` then checking `uriComponent[index + 1] == ':'` seems more straightforward.

File: src/Microsoft.AspNetCore.Http.Abstractions/HostString.cs [hunk 5]
```diff
@@ -38,6 +64,92 @@ public bool HasValue
...
+                    return string.Empty;
+                }
+
+                // normalize the value
+                var uriComponent = ToUriComponent();
+
+                int index;
+                if ((index = uriComponent.IndexOf(']')) >= 0)
+                {
+                    // optimization if uriComponent is exactly the domain
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Http.Abstractions/HostString.cs]: FYI `Substring()` does this optimization for you already. Won't allocate in the whole string case.

File: src/Microsoft.AspNetCore.Http.Abstractions/HostString.cs [hunk 6]
```diff
@@ -38,6 +64,92 @@ public bool HasValue
...
+                    {
+                        // no port
+                        return uriComponent;
+                    }
+                }
+            }
+        }
+
+        /// <summary>
+        /// Returns the value of the port part of the host, or <value>0<code> if none is found.
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Http.Abstractions/HostString.cs]: Why doesn't this return `int?` and `null` in the no port case? That would align better with the new constructor.

**Comment by dougbu** [comment]
:shipit:

File: src/Microsoft.AspNetCore.Http.Abstractions/HostString.cs [hunk 7]
```diff
@@ -25,6 +26,31 @@ public HostString(string value)
         }
 
         /// <summary>
+        /// Creates a new HostString without modification.
+        /// </summary>
+        /// <param name="value">The value should be Unicode rather than punycode. IPv6 addresses must use square braces.</param>
+        /// <param name="port">A positive, greater than 0 value representing the port in the host string.</param>
+        public HostString(string value, int? port)
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Http.Abstractions/HostString.cs]: Don't use a nullable. If there's no port then make them call the other constructor.

File: src/Microsoft.AspNetCore.Http.Abstractions/HostString.cs (hunk 2 above)
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Http.Abstractions/HostString.cs]: ToString InvariantCulture

File: src/Microsoft.AspNetCore.Http.Abstractions/HostString.cs [hunk 8]
```diff
@@ -25,6 +26,31 @@ public HostString(string value)
...
+        {
+            if(port.HasValue && port <= 0)
+            {
+                throw new ArgumentException(Resources.FormatException_PortMustBeGreaterThanZero(nameof(port)), nameof(port));
+            }
+
+            int index;
+            if (value.IndexOf('[') == -1
+                && (index = value.IndexOf(':')) >= 0
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Http.Abstractions/HostString.cs]: how about `value.Count(':') >= 2`?

File: src/Microsoft.AspNetCore.Http.Abstractions/HostString.cs [hunk 9]
```diff
@@ -38,6 +64,92 @@ public bool HasValue
         }
 
         /// <summary>
+        /// Returns the value of the domain part of the host. The port is removed if it was present.
+        /// IPv6 addresses will have brackets added if they are missing.
+        /// </summary>
+        /// <returns></returns>
+        public string Domain
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Http.Abstractions/HostString.cs]: Officially it's called the Host, not the Domain. https://www.w3.org/Protocols/rfc2616/rfc2616-sec14.html#sec14.23

File: src/Microsoft.AspNetCore.Http.Abstractions/HostString.cs [hunk 10]
```diff
@@ -38,6 +64,92 @@ public bool HasValue
...
+        {
+            get
+            {
+                if (string.IsNullOrEmpty(_value))
+                {
+                    return string.Empty;
+                }
+
+                // normalize the value
+                var uriComponent = ToUriComponent();
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Http.Abstractions/HostString.cs]: ToUriComponent returns the punycode form. We'd rather return the unicode form from this property. That's the form you'd get from Value and pass into the constructor.

File: src/Microsoft.AspNetCore.Http.Abstractions/HostString.cs (hunk 6 above)
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Http.Abstractions/HostString.cs]: :+1:

File: src/Microsoft.AspNetCore.Http.Abstractions/HostString.cs (hunk 1 above)
Comment (sebastienros) [inline src/Microsoft.AspNetCore.Http.Abstractions/HostString.cs]: I initially used `domain` like you said, but Chris preferred `value`. 2 vs 1 ;)

File: src/Microsoft.AspNetCore.Http.Abstractions/HostString.cs (hunk 3 above)
Comment (sebastienros) [inline src/Microsoft.AspNetCore.Http.Abstractions/HostString.cs]: I followed an example in `PathString` and I assumed this was the convention. Will do.

File: src/Microsoft.AspNetCore.Http.Abstractions/HostString.cs (hunk 8 above)
Comment (sebastienros) [inline src/Microsoft.AspNetCore.Http.Abstractions/HostString.cs]: I used the same logic as in `ToUriComponent`

File: src/Microsoft.AspNetCore.Http.Abstractions/HostString.cs (hunk 1 above)
Comment (sebastienros) [inline src/Microsoft.AspNetCore.Http.Abstractions/HostString.cs]: Should I use host instead then (based on another comment stating it's not domain but host actually)

Comment (Tratcher) [inline src/Microsoft.AspNetCore.Http.Abstractions/HostString.cs]: Hmm
```
HostString(string value)
HostString(string host, int port)
```

File: src/Microsoft.AspNetCore.Http.Abstractions/HostString.cs (hunk 10 above)
Comment (sebastienros) [inline src/Microsoft.AspNetCore.Http.Abstractions/HostString.cs]: Does it mean we should unencode it if the value contains punycode ?

Comment (Tratcher) [inline src/Microsoft.AspNetCore.Http.Abstractions/HostString.cs]: No, _value is documented to be in it's decoded state.

Comment (Tratcher) [inline src/Microsoft.AspNetCore.Http.Abstractions/HostString.cs]: FromUriComponent takes care of that.

File: src/Microsoft.AspNetCore.Http.Abstractions/HostString.cs [hunk 11]
```diff
@@ -38,6 +64,88 @@ public bool HasValue
...
+                    return _value;
+                }
+            }
+        }
+
+        /// <summary>
+        /// Returns the value of the port part of the host, or <value>0</value> if none is found.
+        /// </summary>
+        /// <returns></returns>
+        public int Port
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Http.Abstractions/HostString.cs]: `int?`

Comment (sebastienros) [inline src/Microsoft.AspNetCore.Http.Abstractions/HostString.cs]: `0` means no port, you prefer `int?` ?

File: src/Microsoft.AspNetCore.Http.Abstractions/HostString.cs [hunk 12]
```diff
@@ -38,6 +64,88 @@ public bool HasValue
...
+                }
+                else
+                {
+                    // Has a port
+                    int.TryParse(_value.Substring(index + 1), out port);
+                }
+
+                return port;
+            }
+        }
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Http.Abstractions/HostString.cs]: There really is a lot of duplicate code here. How about sharing most of it with:
`private void GetParts(out string host, out string port)`

Comment (sebastienros) [inline src/Microsoft.AspNetCore.Http.Abstractions/HostString.cs]: Ok.

File: src/Microsoft.AspNetCore.Http.Abstractions/HostString.cs (hunk 11 above)
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Http.Abstractions/HostString.cs]: Yes, no magic values please. `var port = hostString.Port; if (port.HasValue)` is much better than `if (port != 0)`

**Comment by Tratcher** [comment]
:shipit:

**Pull request closed without merging**